src/backend/sgx/thread.rs

```diff
@@ -160,32 +161,93 @@ fn sgx_enarxcall<'a>(
+            // # Safety
+            //
+            // The device driver checks that the request is within the address
+            // range.
+            std::mem::forget(unsafe {
+                Map::bytes(*length)
+                    .onto(*addr)
+                    .from(&mut keep.enclave.try_clone().unwrap(), 0)
+                    .with_kind(Shared)
+                    .with(perms::Unknown(*prot as i32))
+                    .context("mmap() failed")?
+            });
+
             *ret = 0;
             Ok(None)
         }
 
         item::Enarxcall {
-            num: item::enarxcall::Number::ResetSgxPermissions,
-            argv: [addr, length, ..],
+            num: item::enarxcall::Number::MprotectHost,
+            argv: [addr, length, prot, ..],
             ret,
             ..
         } => {
-            // TODO: Once SGX2 v5 series is out, reset permissions to zero
-            // instead. Resetting  to 'read' is just a workaround for v4
-            // ioctl's.
-            let mut parameters = RestrictPermissions::new(*addr - keep.mem.addr(), *length, 1);
+            let mem_end = keep.mem.addr() + keep.mem.size();
+            let end = *addr + *length;
+
+            // Check that the span is within the enclave address range:
+            if *addr < keep.mem.addr() || end > mem_end {
+                panic!("mprotect() is out of range");
+            }
+
+            // # Safety
+            //
+            // Could change permissions of non-enclave memory, if out of the
+            // enclave address range.
+            if unsafe { libc::mprotect(*addr as _, *length, *prot as i32) } != 0 {
+                panic!("mprotect() failed");
+            }
+
+            // FIXME: Replace with `PROT_READ` with `PROT_NONE`, once migrating
+            // to the mainline kernel.  `PROT_READ` is just a workaround for the
+            // patch set version 4.
+            let mut parameters = RestrictPermissions::new(
+                *addr - keep.mem.addr(),
+                *length,
+                libc::PROT_READ as usize,
+            );
             ENCLAVE_RESTRICT_PERMISSIONS
                 .ioctl(&mut keep.enclave.try_clone().unwrap(), &mut parameters)
                 .context("ENCLAVE_RESTRICT_PERMISSIONS failed")?;
+
+            *ret = 0;
+            Ok(None)
+        }
+
+        item::Enarxcall {
+            num: item::enarxcall::Number::MunmapHost,
+            argv: [addr, length, ..],
+            ret,
+            ..
+        } => {
+            let mem_end = keep.mem.addr() + keep.mem.size();
+            let end = *addr + *length;
+
+            // Check that the span is within the enclave address range:
+            if *addr < keep.mem.addr() || end > mem_end {
+                panic!("munmap() is out of range");
+            }
+
+            // # Safety
+            //
+            // Could unmap non-enclave memory, if out of the enclave address
+            // range.
```

Comment:
```suggestion
            // SAFETY:
            // Parameters have been sanity checked before, that only enclave memory is unmapped.
```
https://rust-lang.github.io/rust-clippy/master/index.html#undocumented_unsafe_blocks